### ISSUE TITLE
Bug: conflicting warnings - array-callback-return and no-unreachable with ts strictNullChecks turned on

### ISSUE THREAD

**Issue by aliak00** [eslint/eslint#16364]
### Environment

Node version: v16.17.0
npm version: v8.15.0
Local ESLint version: Not found
Global ESLint version: Not found
Operating System: darwin 21.6.0

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>
More or less default with create react app 4 I think. Except for:

```
  "eslintConfig": {
    "extends": "react-app",
    "plugins": [
      "eslint-plugin-local-rules"
    ],
    "rules": {
      "curly": [
        "error",
        "all"
      ],
      "no-shadow": "off",
      "@typescript-eslint/no-shadow": [
        "error"
      ],
      "prefer-const": [
        "error",
        {
          "destructuring": "any",
          "ignoreReadBeforeAssign": false
        }
      ],
      "local-rules/disallow-sanitize-html-import": [
        "error"
      ]
    },
```

</details>

```ts
  return featureInfo.audience.some((a): boolean => {
    switch (a) {
      case 'developers':
        return developers.has(user.email);
      case 'internal':
        return isTanaPerson(user.email);
      case 'not-production':
        return process.env.REACT_APP_ENV !== 'production';
      case 'bleeding-edge-production':
        return process.env.REACT_APP_ENV === 'production' && process.env.REACT_APP_RELEASE_CHANNEL === 'be';
      case 'canary':
        return process.env.REACT_APP_ENV === 'production' && process.env.REACT_APP_RELEASE_CHANNEL === 'canary';
      case 'tana-labs':
        return true;
      case 'any':
        return true;
    }
  });
```

### What did you expect to happen?

No warnings

### What actually happened?

Got a warning for `array-callback-return`

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

I added a `return false` at the end of the function, but then the warnings are "unreachable code" - which is correct because of the ts strictNullChecks setting

**Comment by mdjermanovic** [comment]
Hi @aliak00, thanks for the issue!

The core rule `array-callback-return` does not have type information to know that all possible values of `a` are covered in the switch statement. There's an accepted proposal for a new rule in `@typescript-eslint/eslint-plugin` that could be used instead of the core `array-callback-return`: https://github.com/typescript-eslint/typescript-eslint/issues/2841

**Comment by ljharb** [comment]
Why not do
```js
case 'any':
default:
  return true;
```
?

**Comment by sp3ctum** [comment]
@ljharb it may be better to not have a default case because then you can get compilation errors (if your type is updated) using a linter such as this: https://typescript-eslint.io/rules/switch-exhaustiveness-check/
